File: modules/content_service/src/client/content_storage.cpp
#include "oink_judge/content_service/client/content_storage.h"

#include "oink_judge/content_service/client/config_utils.h"
#include "oink_judge/content_service/client/content_service_stub.h"
#include "oink_judge/content_service/config_utils.h"
#include "oink_judge/content_service/manifest_storage.h"

#include <oink_judge/config/common_utils.h>
#include <oink_judge/logger/logger.h>
#include <oink_judge/utils/crypto.h>
#include <oink_judge/utils/filesystem.h>

#include <algorithm>
#include <filesystem>

namespace oink_judge::content_service {

using namespace utils::filesystem;
namespace fs = std::filesystem;

using config::requireHasValue;

auto ContentStorage::instance() -> ContentStorage& {
    static ContentStorage instance;
    return instance;
}

auto ContentStorage::syncContent(std::string content_type, std::string content_id) -> awaitable<void> {
    fs::path content_path = requireHasValue(getContentDirectory(content_type)) / content_id;

    json server_manifest = co_await getManifestFromServer(content_type, content_id);

    std::vector<ContentChange> changes =
        compareManifests(ManifestStorage::instance().getManifest(content_type, content_id), server_manifest);

    if (changes.empty()) {
        co_return;
    }

    logger::logInfo("content_storage",
                    "Syncing " + std::to_string(changes.size()) + " change(s) for " + content_type + "/" + content_id);

    for (const auto& change : changes) {
        if (change.type == ContentChange::Type::ADDED || change.type == ContentChange::Type::MODIFIED) {
            std::string file = co_await getFileFromServer(content_type, content_id, change.file_path);
            storeFile(content_path / change.file_path, file);
            setPermissions(content_path / change.file_path, getPermissionsFromManifest(server_manifest, change.file_path));
        } else if (change.type == ContentChange::Type::ATTRIBUTES_CHANGED) {
            setPermissions(content_path / change.file_path, getPermissionsFromManifest(server_manifest, change.file_path));
        } else if (change.type == ContentChange::Type::REMOVED) {
            removeFileOrDirectory(content_path / change.file_path);
        }
    }
}

auto ContentStorage::syncFile(std::string content_type, std::string content_id, std::string file_path) -> awaitable<void> {
    fs::path content_path = requireHasValue(getContentDirectory(content_type)) / content_id;
    const fs::path target_file_path = fs::path(std::move(file_path));

    json server_manifest = co_await getManifestFromServer(content_type, content_id);

    std::vector<ContentChange> changes =
        compareManifests(ManifestStorage::instance().getManifest(content_type, content_id), server_manifest);

    const auto [removed_begin, removed_end] = std::ranges::remove_if(
        changes, [&](const ContentChange& change) -> bool { return change.file_path != target_file_path; });
    changes.erase(removed_begin, removed_end);

    if (changes.empty()) {
        co_return;
    }

    logger::logInfo("content_storage", "Syncing file " + target_file_path.string() + " for " + content_type + "/" + content_id);

    for (const auto& change : changes) {
        if (change.type == ContentChange::Type::ADDED || change.type == ContentChange::Type::MODIFIED) {
            std::string file = co_await getFileFromServer(content_type, content_id, change.file_path);
            storeFile(content_path / change.file_path, file);
            setPermissions(content_path / change.file_path, getPermissionsFromManifest(server_manifest, change.file_path));
        } else if (change.type == ContentChange::Type::ATTRIBUTES_CHANGED) {
            setPermissions(content_path / change.file_path, getPermissionsFromManifest(server_manifest, change.file_path));
        } else if (change.type == ContentChange::Type::REMOVED) {
            removeFileOrDirectory(content_path / change.file_path);
        }
    }
}

auto ContentStorage::updateContentOnServer(std::string content_type, std::string content_id) -> awaitable<void> {
    logger::logInfo("content_storage", "Updating content on server for " + content_type + "/" + content_id);
    fs::path content_path = requireHasValue(getContentDirectory(content_type)) / content_id;
    if (!std::filesystem::exists(content_path)) {
        throw std::runtime_error("Content path does not exist: " + content_path.string() + " for " + content_type + "/" +
                                 content_id);
    }

    json server_manifest = co_await getManifestFromServer(content_type, content_id);

    const json local_manifest_json = ManifestStorage::instance().getManifest(content_type, content_id).toJson();

    std::vector<ContentChange> changes = compareManifests(server_manifest, local_manifest_json);

    if (changes.empty()) {
        co_return;
    }

    logger::logInfo("content_storage",
                    "Uploading " + std::to_string(changes.size()) + " change(s) for " + content_type + "/" + content_id);

    for (const auto& change : changes) {
        if (change.type == ContentChange::Type::ADDED) {
            std::string file_content = loadFile(content_path / change.file_path);
            co_await createFileOnServer(content_type, content_id, change.file_path, file_content);
            co_await setPermissionsOnServer(
                content_type, content_id, change.file_path,
                static_cast<uint32_t>(getPermissionsFromManifest(local_manifest_json, change.file_path)));
        } else if (change.type == ContentChange::Type::MODIFIED) {
            std::string file_content = loadFile(content_path / change.file_path);
            co_await updateFileOnServer(content_type, content_id, change.file_path, file_content);
            co_await setPermissionsOnServer(
                content_type, content_id, change.file_path,
                static_cast<uint32_t>(getPermissionsFromManifest(local_manifest_json, change.file_path)));
        } else if (change.type == ContentChange::Type::ATTRIBUTES_CHANGED) {
            co_await setPermissionsOnServer(
                content_type, content_id, change.file_path,
                static_cast<uint32_t>(getPermissionsFromManifest(local_manifest_json, change.file_path)));
        } else if (change.type == ContentChange::Type::REMOVED) {
            co_await removeFileOnServer(content_type, content_id, change.file_path);
        }
    }
}

auto ContentStorage::createContent(std::string content_type, std::string content_id) -> awaitable<void> {
    co_await createContentOnServer(content_type, content_id);
    createDirectoryIfNotExists(requireHasValue(getContentDirectory(content_type)) / content_id);
}

auto ContentStorage::listContent(std::string content_type) -> awaitable<std::vector<std::string>> {
    co_return co_await listContentOnServer(content_type);
}

ContentStorage::ContentStorage() {
    auto stub_opt = getContentStorageStub();
    if (!stub_opt.has_value()) {
        throw std::runtime_error("Failed to create ContentServiceStub: optional expected with value");
    }
    stub_ = std::move(*stub_opt);
}

ContentStorage::ContentStorage(std::unique_ptr<ContentServiceStub> stub) : stub_(std::move(stub)) {}

auto ContentStorage::getManifestFromServer(std::string content_type, std::string content_id) -> awaitable<json> {
    auto manifest_exp = co_await stub_->getManifest(content_type, content_id);
    if (!manifest_exp.has_value()) {
        logger::logError("content_storage", "Failed to get manifest from server: " + manifest_exp.error().error_message());
        throw std::runtime_error("Failed to get manifest from server: " + manifest_exp.error().error_message());
    }

    co_return *manifest_exp;
}

auto ContentStorage::getFileFromServer(std::string content_type, std::string content_id, std::string file_path)
    -> awaitable<std::string> {
    auto file_exp = co_await stub_->getFile(content_type, content_id, file_path);
    if (!file_exp.has_value()) {
        throw std::runtime_error("Failed to get file from server: " + file_exp.error().error_message());
    }
    co_return *file_exp;
}

auto ContentStorage::createFileOnServer(std::string content_type, std::string content_id, std::string file_path,
                                        std::string file_content) -> awaitable<void> {
    auto update_exp = co_await stub_->createFile(content_type, content_id, file_path, file_content);
    if (!update_exp.has_value()) {
        throw std::runtime_error("Failed to create file on server: " + update_exp.error().error_message());
    }
}

auto ContentStorage::updateFileOnServer(std::string content_type, std::string content_id, std::string file_path,
                                        std::string file_content) -> awaitable<void> {
    auto update_exp = co_await stub_->updateFile(content_type, content_id, file_path, file_content);
    if (!update_exp.has_value()) {
        throw std::runtime_error("Failed to update file on server: " + update_exp.error().error_message());
    }
}

auto ContentStorage::removeFileOnServer(std::string content_type, std::string content_id, std::string file_path)
    -> awaitable<void> {
    auto delete_exp = co_await stub_->deleteFile(content_type, content_id, file_path);
    if (!delete_exp.has_value()) {
        throw std::runtime_error("Failed to delete file on server: " + delete_exp.error().error_message());
    }
}

auto ContentStorage::setPermissionsOnServer(std::string content_type, std::string content_id, std::string file_path,
                                            uint32_t permissions) -> awaitable<void> {
    auto set_permissions_exp = co_await stub_->setPermissions(content_type, content_id, file_path, permissions);
    if (!set_permissions_exp.has_value()) {
        throw std::runtime_error("Failed to set permissions on server: " + set_permissions_exp.error().error_message());
    }
}

auto ContentStorage::createContentOnServer(std::string content_type, std::string content_id) -> awaitable<void> {
    auto create_content_exp = co_await stub_->createContent(content_type, content_id);
    if (!create_content_exp.has_value()) {
        throw std::runtime_error("Failed to create content on server: " + create_content_exp.error().error_message());
    }
}

auto ContentStorage::listContentOnServer(std::string content_type) -> awaitable<std::vector<std::string>> {
    auto list_content_exp = co_await stub_->listContent(content_type);
    if (!list_content_exp.has_value()) {
        throw std::runtime_error("Failed to list content on server: " + list_content_exp.error().error_message());
    }
    co_return *list_content_exp;
}

} // namespace oink_judge::content_service

File: app/services/content_service/content_storage.py
from typing import Union

import grpc
import grpc.aio
from oink_judge.pybind11_content_service import (
    ContentChangeType,
    ManifestStorage,
    get_content_directory,
    get_permissions_from_manifest,
)
from oink_judge.pybind11_logger import log_info

from app.config import require_has_value
from app.services.content_service.content_manifest import compare_manifests
from app.services.content_service.content_service_stub import (
    ContentServiceStub,
    get_content_storage_stub,
)
from app.utils import (
    create_directory_if_not_exists,
    load_file,
    remove_file_or_directory,
    set_permissions,
    store_file,
)


def _to_bytes(content: Union[str, bytes]) -> bytes:
    if isinstance(content, bytes):
        return content
    return content.encode("latin-1")


def _to_str(content: Union[str, bytes]) -> str:
    if isinstance(content, str):
        return content
    return content.decode("latin-1")


class ContentStorage:
    _instance: "ContentStorage | None" = None

    def __init__(self, stub: ContentServiceStub | None = None) -> None:
        if stub is None:
            stub = require_has_value(get_content_storage_stub())
        self._stub = stub

    @classmethod
    def instance(cls) -> "ContentStorage":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    async def sync_content(self, content_type: str, content_id: str) -> None:
        content_path = (
            require_has_value(get_content_directory(content_type)) / content_id
        )

        server_manifest = await self._get_manifest_from_server(content_type, content_id)

        local_manifest = ManifestStorage.instance().get_manifest(
            content_type, content_id
        )
        changes = compare_manifests(local_manifest, server_manifest)

        if not changes:
            return

        log_info(
            "content_storage",
            f"Syncing {len(changes)} change(s) for {content_type}/{content_id}",
        )

        for change in changes:
            relative_file_path = str(change.file_path)
            file_path = content_path / change.file_path
            if change.type in (ContentChangeType.ADDED, ContentChangeType.MODIFIED):
                file = await self._get_file_from_server(
                    content_type, content_id, relative_file_path
                )
                store_file(file_path, _to_str(file))
                permissions = get_permissions_from_manifest(
                    server_manifest, change.file_path
                )
                set_permissions(file_path, permissions)
            elif change.type == ContentChangeType.ATTRIBUTES_CHANGED:
                permissions = get_permissions_from_manifest(
                    server_manifest, change.file_path
                )
                set_permissions(file_path, permissions)
            elif change.type == ContentChangeType.REMOVED:
                remove_file_or_directory(file_path)

    async def sync_file(
        self, content_type: str, content_id: str, file_path: str
    ) -> None:
        from pathlib import Path

        content_path = (
            require_has_value(get_content_directory(content_type)) / content_id
        )
        target_file_path = Path(file_path)

        server_manifest = await self._get_manifest_from_server(content_type, content_id)

        local_manifest = ManifestStorage.instance().get_manifest(
            content_type, content_id
        )
        changes = [
            change
            for change in compare_manifests(local_manifest, server_manifest)
            if change.file_path == target_file_path
        ]

        if not changes:
            return

        log_info(
            "content_storage",
            f"Syncing file {target_file_path} for {content_type}/{content_id}",
        )

        for change in changes:
            relative_file_path = str(change.file_path)
            local_file_path = content_path / change.file_path
            if change.type in (ContentChangeType.ADDED, ContentChangeType.MODIFIED):
                file = await self._get_file_from_server(
                    content_type, content_id, relative_file_path
                )
                store_file(local_file_path, _to_str(file))
                permissions = get_permissions_from_manifest(
                    server_manifest, change.file_path
                )
                set_permissions(local_file_path, permissions)
            elif change.type == ContentChangeType.ATTRIBUTES_CHANGED:
                permissions = get_permissions_from_manifest(
                    server_manifest, change.file_path
                )
                set_permissions(local_file_path, permissions)
            elif change.type == ContentChangeType.REMOVED:
                remove_file_or_directory(local_file_path)

    async def update_content_on_server(
        self, content_type: str, content_id: str
    ) -> None:
        content_path = (
            require_has_value(get_content_directory(content_type)) / content_id
        )
        if not content_path.exists():
            raise RuntimeError(f"Content path does not exist: {content_path}")

        server_manifest = await self._get_manifest_from_server(content_type, content_id)

        local_manifest = ManifestStorage.instance().get_manifest(
            content_type, content_id
        )
        local_manifest_json = local_manifest.to_json()
        changes = compare_manifests(server_manifest, local_manifest_json)

        if not changes:
            return

        log_info(
            "content_storage",
            f"Uploading {len(changes)} change(s) for {content_type}/{content_id}",
        )

        for change in changes:
            relative_file_path = str(change.file_path)
            file_path = content_path / change.file_path
            if change.type == ContentChangeType.ADDED:
                file_content = _to_bytes(load_file(file_path))
                await self._create_file_on_server(
                    content_type, content_id, relative_file_path, file_content
                )
                await self._set_permissions_on_server(
                    content_type,
                    content_id,
                    relative_file_path,
                    get_permissions_from_manifest(
                        local_manifest_json, change.file_path
                    ),
                )
            elif change.type == ContentChangeType.MODIFIED:
                file_content = _to_bytes(load_file(file_path))
                await self._update_file_on_server(
                    content_type, content_id, relative_file_path, file_content
                )
                await self._set_permissions_on_server(
                    content_type,
                    content_id,
                    relative_file_path,
                    get_permissions_from_manifest(
                        local_manifest_json, change.file_path
                    ),
                )
            elif change.type == ContentChangeType.ATTRIBUTES_CHANGED:
                await self._set_permissions_on_server(
                    content_type,
                    content_id,
                    relative_file_path,
                    get_permissions_from_manifest(
                        local_manifest_json, change.file_path
                    ),
                )
            elif change.type == ContentChangeType.REMOVED:
                await self._remove_file_on_server(
                    content_type, content_id, relative_file_path
                )

    async def create_content(self, content_type: str, content_id: str) -> None:
        await self._create_content_on_server(content_type, content_id)
        create_directory_if_not_exists(
            require_has_value(get_content_directory(content_type)) / content_id
        )

    async def list_content(self, content_type: str) -> list[str]:
        return await self._list_content_on_server(content_type)

    async def _get_manifest_from_server(
        self, content_type: str, content_id: str
    ) -> dict:
        try:
            return await self._stub.get_manifest(content_type, content_id)
        except grpc.aio.AioRpcError as error:
            raise RuntimeError(
                f"Failed to get manifest from server: {error.details()}"
            ) from error

    async def _get_file_from_server(
        self, content_type: str, content_id: str, file_path: str
    ) -> bytes:
        try:
            return await self._stub.get_file(content_type, content_id, file_path)
        except grpc.aio.AioRpcError as error:
            raise RuntimeError(
                f"Failed to get file from server: {error.details()}"
            ) from error

    async def _create_file_on_server(
        self,
        content_type: str,
        content_id: str,
        file_path: str,
        file_content: bytes,
    ) -> None:
        try:
            await self._stub.create_file(
                content_type, content_id, file_path, file_content
            )
        except grpc.aio.AioRpcError as error:
            raise RuntimeError(
                f"Failed to create file on server: {error.details()}"
            ) from error

    async def _update_file_on_server(
        self,
        content_type: str,
        content_id: str,
        file_path: str,
        file_content: bytes,
    ) -> None:
        try:
            await self._stub.update_file(
                content_type, content_id, file_path, file_content
            )
        except grpc.aio.AioRpcError as error:
            raise RuntimeError(
                f"Failed to update file on server: {error.details()}"
            ) from error

    async def _remove_file_on_server(
        self, content_type: str, content_id: str, file_path: str
    ) -> None:
        try:
            await self._stub.delete_file(content_type, content_id, file_path)
        except grpc.aio.AioRpcError as error:
            raise RuntimeError(
                f"Failed to delete file on server: {error.details()}"
            ) from error

    async def _set_permissions_on_server(
        self,
        content_type: str,
        content_id: str,
        file_path: str,
        permissions: int,
    ) -> None:
        try:
            await self._stub.set_permissions(
                content_type, content_id, file_path, int(permissions)
            )
        except grpc.aio.AioRpcError as error:
            raise RuntimeError(
                f"Failed to set permissions on server: {error.details()}"
            ) from error

    async def _create_content_on_server(
        self, content_type: str, content_id: str
    ) -> None:
        try:
            await self._stub.create_content(content_type, content_id)
        except grpc.aio.AioRpcError as error:
            raise RuntimeError(
                f"Failed to create content on server: {error.details()}"
            ) from error

    async def _list_content_on_server(self, content_type: str) -> list[str]:
        try:
            return await self._stub.list_content(content_type)
        except grpc.aio.AioRpcError as error:
            raise RuntimeError(
                f"Failed to list content on server: {error.details()}"
            ) from error
